// @taiga-ui/legacy/esm2022/components/tooltip/tooltip.module.mjs
import { CommonModule } from '@angular/common';
import { NgModule } from '@angular/core';
import { TuiIcon } from '@taiga-ui/core/components/icon';
import { TuiHint } from '@taiga-ui/core/directives/hint';
import { TuiWrapperModule } from '@taiga-ui/legacy/directives';
import { PolymorpheusOutlet } from '@taiga-ui/polymorpheus';
import { TuiTooltipComponent } from './tooltip.component';
import * as i0 from "@angular/core";
import * as i1 from "@taiga-ui/core/directives/hint";
/**
 * @deprecated: drop in v5.0 use {@link TuiTooltip}
 * https://taiga-ui.dev/components/tooltip
 */
class TuiTooltipModule {
    static { this.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "16.2.12", ngImport: i0, type: TuiTooltipModule, deps: [], target: i0.ɵɵFactoryTarget.NgModule }); }
    static { this.ɵmod = i0.ɵɵngDeclareNgModule({ minVersion: "14.0.0", version: "16.2.12", ngImport: i0, type: TuiTooltipModule, declarations: [TuiTooltipComponent], imports: [CommonModule, TuiIcon, PolymorpheusOutlet, i1.TuiHintComponent, i1.TuiHintDirective, i1.TuiHintOptionsDirective, i1.TuiHintUnstyled, i1.TuiHintDriver, i1.TuiHintPosition, i1.TuiHintHover, i1.TuiHintOverflow, i1.TuiHintDescribe, i1.TuiHintHost, i1.TuiHintManual, i1.TuiHintPointer, TuiWrapperModule], exports: [TuiTooltipComponent] }); }
    static { this.ɵinj = i0.ɵɵngDeclareInjector({ minVersion: "12.0.0", version: "16.2.12", ngImport: i0, type: TuiTooltipModule, imports: [CommonModule, TuiIcon, i1.TuiHintComponent, TuiWrapperModule] }); }
}
export { TuiTooltipModule };
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "16.2.12", ngImport: i0, type: TuiTooltipModule, decorators: [{
            type: NgModule,
            args: [{
                    imports: [CommonModule, TuiIcon, PolymorpheusOutlet, ...TuiHint, TuiWrapperModule],
                    declarations: [TuiTooltipComponent],
                    exports: [TuiTooltipComponent],
                }]
        }] });
