// @taiga-ui/addon-charts/esm2022/components/pie-chart/pie-chart.directive.mjs
import { Directive, inject, Input } from '@angular/core';
import { takeUntilDestroyed } from '@angular/core/rxjs-interop';
import { WA_ANIMATION_FRAME, WA_PERFORMANCE } from '@ng-web-apis/common';
import { tuiDescribeSector } from '@taiga-ui/addon-charts/utils';
import { tuiZonefree } from '@taiga-ui/cdk/observables';
import { tuiInjectElement } from '@taiga-ui/cdk/utils/dom';
import { tuiClamp } from '@taiga-ui/cdk/utils/math';
import { tuiEaseInOutQuad } from '@taiga-ui/cdk/utils/miscellaneous';
import { TUI_ANIMATIONS_SPEED } from '@taiga-ui/core/tokens';
import { tuiGetDuration } from '@taiga-ui/core/utils/miscellaneous';
import { BehaviorSubject, map, pairwise, switchMap, takeWhile } from 'rxjs';
import * as i0 from "@angular/core";
class TuiPieChartDirective {
    constructor() {
        this.sector$ = new BehaviorSubject([0, 0]);
        this.el = tuiInjectElement();
        this.performance = inject(WA_PERFORMANCE);
        this.animationFrame$ = inject(WA_ANIMATION_FRAME);
        this.speed = inject(TUI_ANIMATIONS_SPEED);
        this.$ = this.sector$
            .pipe(pairwise(), switchMap(([prev, cur]) => {
            const now = this.performance.now();
            const startDelta = cur[0] - prev[0];
            const endDelta = cur[1] - prev[1];
            return this.animationFrame$.pipe(map((timestamp) => tuiEaseInOutQuad(tuiClamp((timestamp - now) / tuiGetDuration(this.speed), 0, 1))), takeWhile((progress) => progress < 1, true), map((progress) => [
                prev[0] + startDelta * progress,
                cur[1] > 359 ? cur[1] : prev[1] + endDelta * progress,
            ]));
        }), tuiZonefree(), takeUntilDestroyed())
            .subscribe(([start, end]) => this.el.setAttribute('d', tuiDescribeSector(start, end)));
    }
    set tuiPieChart(sector) {
        this.sector$.next(sector);
    }
    static { this.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "16.2.12", ngImport: i0, type: TuiPieChartDirective, deps: [], target: i0.ɵɵFactoryTarget.Directive }); }
    static { this.ɵdir = i0.ɵɵngDeclareDirective({ minVersion: "14.0.0", version: "16.2.12", type: TuiPieChartDirective, isStandalone: true, selector: "path[tuiPieChart]", inputs: { tuiPieChart: "tuiPieChart" }, ngImport: i0 }); }
}
export { TuiPieChartDirective };
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "16.2.12", ngImport: i0, type: TuiPieChartDirective, decorators: [{
            type: Directive,
            args: [{
                    standalone: true,
                    selector: 'path[tuiPieChart]',
                }]
        }], propDecorators: { tuiPieChart: [{
                type: Input
            }] } });
