// @taiga-ui/layout/esm2022/components/navigation/aside-item.directive.mjs
import { Directive, inject } from '@angular/core';
import { toSignal } from '@angular/core/rxjs-interop';
import { RouterLink, RouterLinkActive } from '@angular/router';
import { tuiProvide } from '@taiga-ui/cdk/utils/miscellaneous';
import { TuiButton } from '@taiga-ui/core/components/button';
import { TUI_DATA_LIST_HOST, TuiDataListComponent, } from '@taiga-ui/core/components/data-list';
import { TUI_DROPDOWN_OPTIONS, TuiDropdownDirective, TuiDropdownManual, TuiDropdownPositionSided, } from '@taiga-ui/core/directives/dropdown';
import { TuiIcons } from '@taiga-ui/core/directives/icons';
import { TUI_COMMON_ICONS, TUI_ICON_END } from '@taiga-ui/core/tokens';
import { TUI_CHEVRON, TuiChevron } from '@taiga-ui/kit/directives/chevron';
import { TuiHintAsideDirective } from './hint-aside.directive';
import * as i0 from "@angular/core";
import * as i1 from "./hint-aside.directive";
import * as i2 from "@taiga-ui/core/directives/dropdown";
import * as i3 from "@taiga-ui/core/components/button";
import * as i4 from "@angular/router";
class TuiAsideItemDirective {
    constructor() {
        this.icon = inject(TUI_COMMON_ICONS).check;
        this.icons = inject(TuiIcons);
        this.link = inject(RouterLink, { self: true, optional: true });
        this.active = toSignal(inject(RouterLinkActive).isActiveChange);
        this.size = 's';
    }
    static { this.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "16.2.12", ngImport: i0, type: TuiAsideItemDirective, deps: [], target: i0.ɵɵFactoryTarget.Directive }); }
    static { this.ɵdir = i0.ɵɵngDeclareDirective({ minVersion: "14.0.0", version: "16.2.12", type: TuiAsideItemDirective, isStandalone: true, selector: "[tuiAsideItem]", host: { attributes: { "tuiButton": "" }, properties: { "class._link": "link", "class._active": "active()", "class._custom": "icon !== icons.iconEnd" } }, providers: [
            tuiProvide(TUI_DATA_LIST_HOST, TuiAsideItemDirective),
            provideIcon(),
            provideDropdown(),
        ], hostDirectives: [{ directive: i1.TuiHintAsideDirective }, { directive: i2.TuiDropdownManual }, { directive: i2.TuiDropdownPositionSided }, { directive: i3.TuiButton }, { directive: i4.RouterLinkActive, inputs: ["routerLinkActiveOptions", "routerLinkActiveOptions"] }], ngImport: i0 }); }
}
export { TuiAsideItemDirective };
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "16.2.12", ngImport: i0, type: TuiAsideItemDirective, decorators: [{
            type: Directive,
            args: [{
                    standalone: true,
                    selector: '[tuiAsideItem]',
                    providers: [
                        tuiProvide(TUI_DATA_LIST_HOST, TuiAsideItemDirective),
                        provideIcon(),
                        provideDropdown(),
                    ],
                    hostDirectives: [
                        TuiHintAsideDirective,
                        TuiDropdownManual,
                        TuiDropdownPositionSided,
                        TuiButton,
                        {
                            directive: RouterLinkActive,
                            inputs: ['routerLinkActiveOptions'],
                        },
                    ],
                    host: {
                        tuiButton: '',
                        '[class._link]': 'link',
                        '[class._active]': 'active()',
                        '[class._custom]': 'icon !== icons.iconEnd',
                    },
                }]
        }] });
function provideIcon() {
    return {
        provide: TUI_ICON_END,
        useFactory: () => {
            if (inject(TuiChevron, { optional: true, self: true })) {
                return inject(TUI_CHEVRON);
            }
            const { check, more } = inject(TUI_COMMON_ICONS);
            const active = inject(TuiDataListComponent, { optional: true }) &&
                inject(RouterLink, { self: true, optional: true })
                ? check
                : '';
            return inject(TuiDropdownDirective, { self: true, optional: true })
                ? more
                : active;
        },
    };
}
function provideDropdown() {
    return {
        provide: TUI_DROPDOWN_OPTIONS,
        useFactory: () => ({
            ...inject(TUI_DROPDOWN_OPTIONS, { skipSelf: true }),
            offset: inject(TuiDataListComponent, { optional: true }) ? 8 : 12,
        }),
    };
}
