// @taiga-ui/layout/esm2022/components/navigation/aside-group.component.mjs
import { ChangeDetectionStrategy, Component, computed, ContentChild, inject, Input, Output, signal, TemplateRef, ViewChild, ViewEncapsulation, } from '@angular/core';
import { toObservable } from '@angular/core/rxjs-interop';
import { tuiDirectiveBinding, tuiProvide } from '@taiga-ui/cdk/utils/miscellaneous';
import { TUI_DATA_LIST_HOST, TuiDataList } from '@taiga-ui/core/components/data-list';
import { TuiExpand } from '@taiga-ui/core/components/expand';
import { TuiDropdownDirective, TuiDropdownHover, TuiDropdownOpen, TuiDropdownPositionSided, } from '@taiga-ui/core/directives/dropdown';
import { TuiChevron } from '@taiga-ui/kit/directives/chevron';
import { PolymorpheusOutlet } from '@taiga-ui/polymorpheus';
import { skip } from 'rxjs';
import { TuiAsideComponent } from './aside.component';
import * as i0 from "@angular/core";
import * as i1 from "@taiga-ui/core/directives/dropdown";
import * as i2 from "@taiga-ui/core/components/data-list";
import * as i3 from "@taiga-ui/core/components/expand";
class TuiAsideGroupComponent {
    constructor() {
        this.aside = inject(TuiAsideComponent);
        this.open = signal(false);
        this.template = null;
        this.expanded = computed(() => this.aside.expanded() && this.open());
        this.binding = tuiDirectiveBinding(TuiDropdownDirective, 'tuiDropdown', computed(() => (this.aside.expanded() ? null : this.datalist)));
        this.openChange = toObservable(this.open).pipe(skip(1));
        this.size = 's';
    }
    set openSetter(open) {
        this.toggle(open);
    }
    toggle(open = !this.open()) {
        this.open.set(open && this.aside.expanded());
        this.chevron?.chevron.set(this.open());
    }
    static { this.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "16.2.12", ngImport: i0, type: TuiAsideGroupComponent, deps: [], target: i0.ɵɵFactoryTarget.Component }); }
    static { this.ɵcmp = i0.ɵɵngDeclareComponent({ minVersion: "14.0.0", version: "16.2.12", type: TuiAsideGroupComponent, isStandalone: true, selector: "tui-aside-group", inputs: { openSetter: ["open", "openSetter"] }, outputs: { openChange: "openChange" }, providers: [tuiProvide(TUI_DATA_LIST_HOST, TuiAsideGroupComponent)], queries: [{ propertyName: "chevron", first: true, predicate: TuiChevron, descendants: true, static: true }, { propertyName: "template", first: true, predicate: TemplateRef, descendants: true }], viewQueries: [{ propertyName: "datalist", first: true, predicate: ["datalist"], descendants: true, static: true }], hostDirectives: [{ directive: i1.TuiDropdownDirective }, { directive: i1.TuiDropdownHover }, { directive: i1.TuiDropdownPositionSided }, { directive: i1.TuiDropdownOpen }], ngImport: i0, template: "<div\n    #host\n    (click)=\"toggle()\"\n>\n    <ng-content />\n</div>\n<tui-expand [expanded]=\"expanded()\">\n    <ng-container *polymorpheusOutlet=\"template\" />\n</tui-expand>\n<ng-template #datalist>\n    <tui-data-list>\n        <tui-opt-group [label]=\"host.textContent\">\n            <ng-container *polymorpheusOutlet=\"template\" />\n        </tui-opt-group>\n    </tui-data-list>\n</ng-template>\n", dependencies: [{ kind: "directive", type: PolymorpheusOutlet, selector: "[polymorpheusOutlet]", inputs: ["polymorpheusOutlet", "polymorpheusOutletContext"] }, { kind: "component", type: i2.TuiDataListComponent, selector: "tui-data-list", inputs: ["emptyContent", "size"] }, { kind: "directive", type: i2.TuiOptGroup, selector: "tui-opt-group", inputs: ["label"] }, { kind: "component", type: i3.TuiExpandComponent, selector: "tui-expand", inputs: ["async", "expanded"] }], changeDetection: i0.ChangeDetectionStrategy.OnPush, encapsulation: i0.ViewEncapsulation.None }); }
}
export { TuiAsideGroupComponent };
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "16.2.12", ngImport: i0, type: TuiAsideGroupComponent, decorators: [{
            type: Component,
            args: [{ standalone: true, selector: 'tui-aside-group', imports: [PolymorpheusOutlet, TuiDataList, TuiExpand], encapsulation: ViewEncapsulation.None, changeDetection: ChangeDetectionStrategy.OnPush, providers: [tuiProvide(TUI_DATA_LIST_HOST, TuiAsideGroupComponent)], hostDirectives: [
                        TuiDropdownDirective,
                        TuiDropdownHover,
                        TuiDropdownPositionSided,
                        TuiDropdownOpen,
                    ], template: "<div\n    #host\n    (click)=\"toggle()\"\n>\n    <ng-content />\n</div>\n<tui-expand [expanded]=\"expanded()\">\n    <ng-container *polymorpheusOutlet=\"template\" />\n</tui-expand>\n<ng-template #datalist>\n    <tui-data-list>\n        <tui-opt-group [label]=\"host.textContent\">\n            <ng-container *polymorpheusOutlet=\"template\" />\n        </tui-opt-group>\n    </tui-data-list>\n</ng-template>\n" }]
        }], propDecorators: { datalist: [{
                type: ViewChild,
                args: ['datalist', { static: true }]
            }], chevron: [{
                type: ContentChild,
                args: [TuiChevron, { static: true }]
            }], template: [{
                type: ContentChild,
                args: [TemplateRef]
            }], openChange: [{
                type: Output
            }], openSetter: [{
                type: Input,
                args: ['open']
            }] } });
