// @taiga-ui/addon-table/esm2022/directives/table-filters/generic-filter.directive.mjs
import { Directive, Input } from '@angular/core';
import { TUI_TRUE_HANDLER } from '@taiga-ui/cdk/constants';
import { tuiProvide } from '@taiga-ui/cdk/utils/miscellaneous';
import { AbstractTuiTableFilter } from './abstract-table-filter';
import * as i0 from "@angular/core";
class TuiGenericFilter extends AbstractTuiTableFilter {
    constructor() {
        super(...arguments);
        this.filter = TUI_TRUE_HANDLER;
    }
    static { this.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "16.2.12", ngImport: i0, type: TuiGenericFilter, deps: null, target: i0.ɵɵFactoryTarget.Directive }); }
    static { this.ɵdir = i0.ɵɵngDeclareDirective({ minVersion: "14.0.0", version: "16.2.12", type: TuiGenericFilter, isStandalone: true, selector: "[tuiGenericFilter]", inputs: { filter: ["tuiGenericFilter", "filter"] }, providers: [tuiProvide(AbstractTuiTableFilter, TuiGenericFilter)], usesInheritance: true, ngImport: i0 }); }
}
export { TuiGenericFilter };
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "16.2.12", ngImport: i0, type: TuiGenericFilter, decorators: [{
            type: Directive,
            args: [{
                    standalone: true,
                    selector: '[tuiGenericFilter]',
                    providers: [tuiProvide(AbstractTuiTableFilter, TuiGenericFilter)],
                }]
        }], propDecorators: { filter: [{
                type: Input,
                args: ['tuiGenericFilter']
            }] } });
