// @taiga-ui/addon-table/esm2022/directives/table-filters/table-filter.directive.mjs
import { Directive, inject, Input } from '@angular/core';
import { NgControl } from '@angular/forms';
import { TuiTableHead } from '@taiga-ui/addon-table/components';
import { defer, distinctUntilChanged, EMPTY, merge } from 'rxjs';
import { AbstractTuiTableFilter } from './abstract-table-filter';
import { TuiTableFiltersDirective } from './table-filters.directive';
import * as i0 from "@angular/core";
class TuiTableFilterDirective {
    constructor() {
        this.head = inject((TuiTableHead), { optional: true });
        this.delegate = inject((AbstractTuiTableFilter));
        this.control = inject(NgControl);
        this.filters = inject((TuiTableFiltersDirective));
        this.refresh$ = defer(() => merge(this.control.valueChanges || EMPTY, this.control.statusChanges?.pipe(distinctUntilChanged()) || EMPTY));
    }
    ngOnInit() {
        this.filters.register(this);
    }
    ngOnDestroy() {
        this.filters.unregister(this);
    }
    filter(item) {
        const { disabled, value } = this.control;
        return (!!disabled ||
            !this.key ||
            this.delegate.filter(item[this.key], value));
    }
    get key() {
        return this.tuiTableFilter || this.head?.tuiHead;
    }
    static { this.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "16.2.12", ngImport: i0, type: TuiTableFilterDirective, deps: [], target: i0.ɵɵFactoryTarget.Directive }); }
    static { this.ɵdir = i0.ɵɵngDeclareDirective({ minVersion: "14.0.0", version: "16.2.12", type: TuiTableFilterDirective, isStandalone: true, selector: "[tuiTableFilter]", inputs: { tuiTableFilter: "tuiTableFilter" }, ngImport: i0 }); }
}
export { TuiTableFilterDirective };
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "16.2.12", ngImport: i0, type: TuiTableFilterDirective, decorators: [{
            type: Directive,
            args: [{
                    standalone: true,
                    selector: '[tuiTableFilter]',
                }]
        }], propDecorators: { tuiTableFilter: [{
                type: Input
            }] } });
