// @taiga-ui/addon-table/esm2022/components/table/directives/head.directive.mjs
import { Directive, inject, Input, TemplateRef } from '@angular/core';
import * as i0 from "@angular/core";
class TuiTableHead {
    constructor() {
        this.tuiHead = '';
        this.template = inject((TemplateRef));
    }
    static { this.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "16.2.12", ngImport: i0, type: TuiTableHead, deps: [], target: i0.ɵɵFactoryTarget.Directive }); }
    static { this.ɵdir = i0.ɵɵngDeclareDirective({ minVersion: "14.0.0", version: "16.2.12", type: TuiTableHead, isStandalone: true, selector: "[tuiHead]", inputs: { tuiHead: "tuiHead" }, ngImport: i0 }); }
}
export { TuiTableHead };
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "16.2.12", ngImport: i0, type: TuiTableHead, decorators: [{
            type: Directive,
            args: [{
                    standalone: true,
                    selector: '[tuiHead]',
                }]
        }], propDecorators: { tuiHead: [{
                type: Input,
                args: [{ required: true }]
            }] } });
